### PR TITLE
update OptParse to work on 2and3

### PR DESCRIPTION
**Pull request by ethanhs**
Continued from #1026, this is a much better approach to the stub file. Huge thanks to Jelle for excellent and thorough feedback on my last stub.

### CODE REVIEW

File: stdlib/2and3/optparse.pyi [hunk 1]
```diff
@@ -0,0 +1,226 @@
+# Generated by pytype, with only minor tweaks. Might be incomplete.
+import sys
+from typing import Any, Optional, List, Callable, Tuple, Dict, Iterable, Union, Mapping, IO
+
+# See https://groups.google.com/forum/#!topic/python-ideas/gA1gdj3RZ5g
+if sys.version_info >= (3,):
+    Text = str
+else:
+    Text = Union[str, unicode]
```
Comment (matthiaskramm) [inline stdlib/2and3/optparse.pyi]: If you do have to define your own type, I'd prefer if you gave it a new name. There's already a `typing.Text`, so this is confusing. `OptText`? `_Text`? `AnyText`?

File: stdlib/2and3/optparse.pyi [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+    def __init__(self, indent_increment: int, max_help_position: int, width: Optional[int], short_first: int) -> None: ...
+    def _format_text(self, text: Text) -> Text: ...
+    def dedent(self) -> None: ...
+    def expand_default(self, option: Option) -> Text: ...
+    def format_description(self, description: Text) -> Text: ...
+    def format_epilog(self, epilog) -> Text: ...
+    def format_heading(self, heading: Any) -> Text: ...
+    def format_option(self, option: OptionParser) -> Text: ...
+    def format_option_strings(self, option: OptionParser) -> Any: ...
+    def format_usage(self, usage: Any) -> Text: ...
```
Comment (matthiaskramm) [inline stdlib/2and3/optparse.pyi]: Some of these seem to be abstract.

File: stdlib/2and3/optparse.pyi [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+    title = ...  # type: Text
+    def __init__(self, parser: OptionParser, title: Text, description: Optional[Text]) -> None: ...
+    def _create_option_list(self) -> None: ...
+    def set_title(self, title: Text) -> None: ...
+
+class OptionParser(OptionContainer):
+    allow_interspersed_args = ...  # type: bool
+    epilog = ...  # type: Any
+    formatter = ...  # type: HelpFormatter
+    largs = ...  # type: Union[None, List[Text]]
```
Comment (matthiaskramm) [inline stdlib/2and3/optparse.pyi]: `Optional[List[Text]]`

File: stdlib/2and3/optparse.pyi (hunk 1 above)
Comment (ethanhs) [inline stdlib/2and3/optparse.pyi]: Ah, this is true. Looking at argparse, I'll change it to `_Text`

File: stdlib/2and3/optparse.pyi (hunk 2 above)
Comment (ethanhs) [inline stdlib/2and3/optparse.pyi]: Several of these are abstract. I don't entirely understand why that is significant? They aren't decorated as `abstractmethod`.

**Comment by ethanhs** [comment]
I responded and resolved to @matthiaskramm's reviews. I didn't do anything about his comments about various functions being abstract, as I didn't see a problem with the stub for those. If I am missing something, please let me know.

File: stdlib/2and3/optparse.pyi (hunk 2 above)
Comment (matthiaskramm) [inline stdlib/2and3/optparse.pyi]: Which means you can subclass HelpFormatter without overriding the "abstract" methods, but you'll get an error once you call the method. I wonder whether type checkers should warn about that.

**Pull request merged**